### PR TITLE
Fix handling path with spaces

### PR DESCRIPTION
**Pull request by Fuchs4**
This fixes handling path with spaces in them.
av1an stores commands in strings, which are split before passing them to the subproces module. This causes any path with spaces in it to be split in multiple parts.
This patch changes all commands (i hope i found all) to be stored in tupel, which can be directly passed to the subprocess module, eliminating the need for splitting them.
For concat i have used shlex to quote paths, which can quote/escape most (if not all) unicode charactes correctly so that ffmpeg sees the correct path.

### CODE REVIEW

**Comment by n9Mtq4** [comment]
Might it be possible to achieve the same functionality by using `shlex.split` and keeping the commands as strings internally?

File: Av1an/bar.py [hunk 1]
```diff
@@ -34,13 +34,11 @@ def update(self, value):
 BaseManager.register('Counter', Counter)
 
 
-def make_pipes(ffmpeg_gen, command):
+def make_pipes(ffmpeg_gen_cmd: list, command: list):
```
Comment (n9Mtq4) [inline Av1an/bar.py:37]: `ffmpeg_gen_cmd` is now a `Tuple[str]`.

`command` is now a tuple containing a list and a tuple `(['ffmpeg', ...], ('aomenc', ...))`. `command` should contain only 1 type inside of it. Pick either `Tuple[List[str]]` or `Tuple[Tuple[str]]`. Since it seems like most commands are going to be tuples, I would prefer `Tuple[Tuple[str]]`.

```suggestion
def make_pipes(ffmpeg_gen_cmd: Tuple[str], command: Tuple[Tuple[str]]):
```

File: Av1an/bar.py [hunk 2]
```diff
@@ -64,7 +62,7 @@ def process_pipe(pipe):
 
 
 def make_vvc_pipe(command):
-    pipe = subprocess.Popen(command.split(), stdout=PIPE,
+    pipe = subprocess.Popen(command[0], stdout=PIPE,
```
Comment (n9Mtq4) [inline Av1an/bar.py:65]: `command` here was only 1 string and is now only 1 tuple, so no indexing is needed.

```suggestion
    pipe = subprocess.Popen(command, stdout=PIPE,
```

File: Av1an/chunk_queue.py [hunk 3]
```diff
@@ -127,7 +127,7 @@ def create_vsffms2_chunk(args: Args, index: int, load_script: Path, frame_start:
     frames = frame_end - frame_start
     frame_end -= 1  # the frame end boundary is actually a frame that should be included in the next chunk
 
-    ffmpeg_gen_cmd = f'vspipe {load_script} -y - -s {frame_start} -e {frame_end}'
+    ffmpeg_gen_cmd = ('vspipe', load_script, '-y', '-', '-s', frame_start, '-e', frame_end)
```
Comment (n9Mtq4) [inline Av1an/chunk_queue.py:130]: Some of these are not strings and now need to be explicitly converted to strings.

```suggestion
    ffmpeg_gen_cmd = ('vspipe', load_script.as_posix(), '-y', '-', '-s', str(frame_start), '-e', str(frame_end))
```

File: Av1an/chunk_queue.py [hunk 4]
```diff
@@ -147,7 +147,10 @@ def create_video_queue_select(args: Args, split_locations: List[int]) -> List[Ch
     """
     # add first frame and last frame
     last_frame = frame_probe(args.input)
-    split_locs_fl = [0] + split_locations + [last_frame]
+    if 0 not in split_locations:
+        split_locs_fl = [0] + split_locations
+    if last_frame not in split_locations:
+        split_locations = split_locations + [last_frame]
```
Comment (n9Mtq4) [inline Av1an/chunk_queue.py:153]: Why is this needed? `split_location` will never have the first or last frame included.

File: Av1an/compose.py [hunk 5]
```diff
@@ -9,7 +9,7 @@
 from .vvc import get_yuv_file_path
 
 
-def gen_pass_commands(args: Args, chunk: Chunk) -> List[str]:
+def gen_pass_commands(args: Args, chunk: Chunk) -> List[List]:
```
Comment (n9Mtq4) [inline Av1an/compose.py:12]: This return type is wrong.
```suggestion
def gen_pass_commands(args: Args, chunk: Chunk) -> Tuple[Tuple[str]]:
```

File: Av1an/chunk.py [hunk 6]
```diff
@@ -11,7 +11,7 @@ class Chunk:
     to be run on this chunk.
     """
 
-    def __init__(self, temp: Path, index: int, ffmpeg_gen_cmd: str, output_ext: str, size: int, frames: int):
+    def __init__(self, temp: Path, index: int, ffmpeg_gen_cmd: list, output_ext: str, size: int, frames: int):
```
Comment (n9Mtq4) [inline Av1an/chunk.py:14]:
```suggestion
    def __init__(self, temp: Path, index: int, ffmpeg_gen_cmd: Tuple[str], output_ext: str, size: int, frames: int):
```

File: Av1an/compose.py [hunk 7]
```diff
@@ -86,9 +86,9 @@ def compose_ffmpeg_pipe(a: Args) -> str:
     Gets the ffmpeg command with filters that pipes into the encoder
 
     :param a: the Args
-    :return: the ffmpeg command as a string
+    :return: the ffmpeg command as a list
     """
-    return f'ffmpeg -y -hide_banner -loglevel error -i - {a.ffmpeg} {a.pix_format} -bufsize 50000K -f yuv4mpegpipe -'
+    return ['ffmpeg', '-y', '-hide_banner', '-loglevel', 'error', '-i', '-', *a.ffmpeg_pipe]
```
Comment (n9Mtq4) [inline Av1an/compose.py:91]: Since every other command is a tuple, this one should be too. You should also update the return type in the function signature.
```suggestion
    return ('ffmpeg', '-y', '-hide_banner', '-loglevel', 'error', '-i', '-', *a.ffmpeg_pipe)
```

File: Av1an/compose.py [hunk 8]
```diff
@@ -86,9 +86,9 @@ def compose_ffmpeg_pipe(a: Args) -> str:
...
 
 
 def compose_svt_av1(a: Args, c: Chunk) -> List[str]:
```
Comment (n9Mtq4) [inline Av1an/compose.py:94]: Update signature.
```suggestion
def compose_svt_av1(a: Args, c: Chunk) -> Tuple[Tuple[str]]:
```

File: Av1an/compose.py [hunk 9]
```diff
@@ -100,14 +100,14 @@ def compose_svt_av1(a: Args, c: Chunk) -> List[str]:
...
-            f'{compose_ffmpeg_pipe(a)} | SvtAv1EncApp -i stdin {a.video_params} -output-stat-file {c.fpf}.stat -b {os.devnull} -',
-            f'{compose_ffmpeg_pipe(a)} | SvtAv1EncApp -i stdin {a.video_params} -input-stat-file {c.fpf}.stat -b {c.output} -',
-        ]
+        return (
+            (compose_ffmpeg_pipe(a), ('SvtAv1EncApp', '-i', 'stdin', *a.video_params, '-output-stat-file', f'{c.fpf}.stat', '-b', os.devnull, '-')),
+            (compose_ffmpeg_pipe(a), ('SvtAv1EncApp', '-i', 'stdin', *a.video_params, '-input-stat-file', f'{c.fpf}.stat', '-b', c.output, '-'))
+        )
 
 
 def compose_rav1e(a: Args, c: Chunk) -> List[str]:
```
Comment (n9Mtq4) [inline Av1an/compose.py:113]:
```suggestion
def compose_rav1e(a: Args, c: Chunk) -> Tuple[Tuple[str]]:
```

File: Av1an/compose.py [hunk 10]
```diff
@@ -120,14 +120,14 @@ def compose_rav1e(a: Args, c: Chunk) -> List[str]:
...
-            f'{compose_ffmpeg_pipe(a)} | rav1e - --first-pass {c.fpf}.stat {a.video_params} --output {os.devnull}',
-            f'{compose_ffmpeg_pipe(a)} | rav1e - --second-pass {c.fpf}.stat {a.video_params} --output {c.output}',
-        ]
+        return (
+            (compose_ffmpeg_pipe(a), ('rav1e', '-', '--first-pass', f'{c.fpf}.stat', *a.video_params, '--output', os.devnull)),
+            (compose_ffmpeg_pipe(a), ('rav1e', '-', '--second-pass', f'{c.fpf}.stat', *a.video_params, '--output', c.output))
+        )
 
 
 def compose_aom(a: Args, c: Chunk) -> List[str]:
```
Comment (n9Mtq4) [inline Av1an/compose.py:133]:
```suggestion
def compose_aom(a: Args, c: Chunk) -> Tuple[Tuple[str]]:
```

File: Av1an/compose.py [hunk 11]
```diff
@@ -139,14 +139,14 @@ def compose_aom(a: Args, c: Chunk) -> List[str]:
...
-            f'{compose_ffmpeg_pipe(a)} | aomenc --passes=2 --pass=1 {a.video_params} --fpf={c.fpf}.log -o {os.devnull} -',
-            f'{compose_ffmpeg_pipe(a)} | aomenc --passes=2 --pass=2 {a.video_params} --fpf={c.fpf}.log -o {c.output} -',
-        ]
+        return (
+            (compose_ffmpeg_pipe(a), ('aomenc', '--passes=2', '--pass=1', *a.video_params, f'--fpf={c.fpf}.log', '-o', os.devnull, '-')),
+            (compose_ffmpeg_pipe(a), ('aomenc', '--passes=2', '--pass=2', *a.video_params, f'--fpf={c.fpf}.log', '-o', c.output, '-'))
+        )
 
 
 def compose_vpx(a: Args, c: Chunk) -> List[str]:
```
Comment (n9Mtq4) [inline Av1an/compose.py:152]:
```suggestion
def compose_vpx(a: Args, c: Chunk) -> Tuple[Tuple[str]]:
```

File: Av1an/compose.py [hunk 12]
```diff
@@ -158,14 +158,14 @@ def compose_vpx(a: Args, c: Chunk) -> List[str]:
...
-            f'{compose_ffmpeg_pipe(a)} | vpxenc --passes=2 --pass=1 {a.video_params} --fpf={c.fpf} -o {os.devnull} -',
-            f'{compose_ffmpeg_pipe(a)} | vpxenc --passes=2 --pass=2 {a.video_params} --fpf={c.fpf} -o {c.output} -',
-        ]
+        return (
+            (compose_ffmpeg_pipe(a), ('vpxenc', '--passes=2', '--pass=1', *a.video_params, f'--fpf={c.fpf}', '-o', os.devnull, '-')),
+            (compose_ffmpeg_pipe(a), ('vpxenc', '--passes=2', '--pass=2', *a.video_params, f'--fpf={c.fpf}', '-o', c.output, '-'))
+        )
 
 
 def compose_x265(a: Args, c: Chunk) -> List[str]:
```
Comment (n9Mtq4) [inline Av1an/compose.py:171]:
```suggestion
def compose_x265(a: Args, c: Chunk) -> Tuple[Tuple[str]]:
```

File: Av1an/compose.py [hunk 13]
```diff
@@ -177,14 +177,14 @@ def compose_x265(a: Args, c: Chunk) -> List[str]:
...
-            f'{compose_ffmpeg_pipe(a)} | x265 --log-level error --pass 1 --y4m {a.video_params} --stats {c.fpf}.log - -o {os.devnull}',
-            f'{compose_ffmpeg_pipe(a)} | x265 --log-level error --pass 2 --y4m {a.video_params} --stats {c.fpf}.log - -o {c.output}',
-        ]
+        return (
+            (compose_ffmpeg_pipe(a), ('x265', '--log-level', 'error', '--pass', '1', '--y4m', *a.video_params, '--stats', f'{c.fpf}.log', '-', '-o', os.devnull)),
+            (compose_ffmpeg_pipe(a), ('x265', '--log-level', 'error', '--pass', '2', '--y4m', *a.video_params, '--stats', f'{c.fpf}.log', '-', '-o', c.output))
+        )
 
 
 def compose_x264(a: Args, c: Chunk) -> List[str]:
```
Comment (n9Mtq4) [inline Av1an/compose.py:190]:
```suggestion
def compose_x264(a: Args, c: Chunk) -> Tuple[Tuple[str]]:
```

File: Av1an/compose.py [hunk 14]
```diff
@@ -196,14 +196,14 @@ def compose_x264(a: Args, c: Chunk) -> List[str]:
...
-            f'{compose_ffmpeg_pipe(a)} | x264 --stitchable --log-level error --pass 1 --demuxer y4m {a.video_params} - --stats {c.fpf}.log - -o {os.devnull}',
-            f'{compose_ffmpeg_pipe(a)} | x264 --stitchable --log-level error --pass 2 --demuxer y4m {a.video_params} - --stats {c.fpf}.log - -o {c.output}',
-        ]
+        return (
+            (compose_ffmpeg_pipe(a), ('x264', '--stitchable', '--log-level', 'error', '--pass', '1', '--demuxer', 'y4m', *a.video_params, '-', '--stats', f'{c.fpf}.log', '-', '-o', os.devnull)),
+            (compose_ffmpeg_pipe(a), ('x264', '--stitchable', '--log-level', 'error', '--pass', '2', '--demuxer', 'y4m', *a.video_params, '-', '--stats', f'{c.fpf}.log', '-', '-o', c.output))
+        )
 
 
 def compose_vvc(a: Args, c: Chunk) -> List[str]:
```
Comment (n9Mtq4) [inline Av1an/compose.py:209]:
```suggestion
def compose_vvc(a: Args, c: Chunk) -> Tuple[Tuple[str]]:
```

File: Av1an/compose.py [hunk 15]
```diff
@@ -215,9 +215,9 @@ def compose_vvc(a: Args, c: Chunk) -> List[str]:
     :return: A list of commands
     """
     yuv_file = get_yuv_file_path(c).as_posix()
-    return [
-        f'vvc_encoder -c {a.vvc_conf} -i {yuv_file} {a.video_params} -f {c.frames} --InputBitDepth=10 --OutputBitDepth=10 -b {c.output}',
-    ]
+    return (
+        ('vvc_encoder', '-c', a.vvc_conf, '-i', yuv_file, *a.video_params, '-f', c.frames, '--InputBitDepth=10', '--OutputBitDepth=10', '-b', c.output),
```
Comment (n9Mtq4) [inline Av1an/compose.py:219]: Need an explicit string conversion.
```suggestion
        ('vvc_encoder', '-c', a.vvc_conf, '-i', yuv_file, *a.video_params, '-f', str(c.frames), '--InputBitDepth=10', '--OutputBitDepth=10', '-b', c.output),
```

File: Av1an/utils.py [hunk 16]
```diff
@@ -45,25 +45,31 @@ def get_cq(command):
     return int(matches[-1])
 
 
-def man_q(command: str, q: int):
+def man_q(command: list, q: int):
```
Comment (n9Mtq4) [inline Av1an/utils.py:48]:
```suggestion
def man_q(command: Tuple[str], q: int):
```

File: Av1an/concat.py [hunk 17]
```diff
@@ -60,26 +61,24 @@ def concatenate_video(temp: Path, output, encoder: str):
...
-        audio = ''
+        audio = ()
 
     if encoder == 'x265':
 
-        cmd = f' ffmpeg -y -fflags +genpts  -hide_banner -loglevel error -f concat -safe 0 -i {temp / "concat"} ' \
-            f'{audio} -c copy -movflags frag_keyframe+empty_moov -map 0  -f mp4 - | ffmpeg -y -hide_banner -loglevel error -i - -c copy {output} '
-        concat = subprocess.run(cmd, shell=True, stdout=PIPE, stderr=STDOUT).stdout
+        cmd = ('ffmpeg', '-y', '-fflags', '+genpts', '-hide_banner', '-loglevel', 'error', '-f', 'concat', '-safe', '0', '-i', temp / "concat", *audio, '-c', 'copy', '-movflags', 'frag_keyframe+empty_moov', '-map', '0', '-f', 'mp4', output)
+        concat = subprocess.run(cmd, shell=False, stdout=PIPE, stderr=STDOUT).stdout
```
Comment (n9Mtq4) [inline Av1an/concat.py:76]: `shell=False` is the default, so just remove it.
```suggestion
        concat = subprocess.run(cmd, stdout=PIPE, stderr=STDOUT).stdout
```

File: Av1an/concat.py [hunk 18]
```diff
@@ -60,26 +61,24 @@ def concatenate_video(temp: Path, output, encoder: str):
...
 
     else:
-        cmd = f' ffmpeg -y -hide_banner -loglevel error -f concat -safe 0 -i {temp / "concat"} ' \
-            f'{audio} -c copy -map 0  -y "{output}"'
+        cmd = ('ffmpeg', '-y', '-hide_banner', '-loglevel', 'error', '-f', 'concat', '-safe', '0', '-i', temp / "concat", *audio, '-c', 'copy', '-map', '0', output)
 
-        concat = subprocess.run(cmd, shell=True, stdout=PIPE, stderr=STDOUT).stdout
+        concat = subprocess.run(cmd, shell=False, stdout=PIPE, stderr=STDOUT).stdout
```
Comment (n9Mtq4) [inline Av1an/concat.py:81]:
```suggestion
        concat = subprocess.run(cmd, stdout=PIPE, stderr=STDOUT).stdout
```

File: Av1an/vmaf.py [hunk 19]
```diff
@@ -54,20 +56,22 @@ def call_vmaf(chunk: Chunk, encoded: Path, n_threads, model, res, fl_path: Path
...
-    vmaf_filter = f"[distorted][ref]libvmaf=log_fmt='json':eof_action=endall:log_path={fl}{mod}{n_threads}\" -f null - "
+    cmd_out = ('-f', 'null', '-')
 
-    cmd = cmd + filter_complex + distorted + ref + vmaf_filter
+    cmd = (*cmd_in, *filter_complex, distorted + ref + vmaf_filter, *cmd_out)
 
-    ffmpeg_gen_pipe = subprocess.Popen(chunk.ffmpeg_gen_cmd.split(), stdout=PIPE, stderr=STDOUT)
-    pipe = subprocess.Popen(cmd, stdin=ffmpeg_gen_pipe.stdout, stdout=PIPE, stderr=STDOUT, shell=True, universal_newlines=True)
+    ffmpeg_gen_pipe = subprocess.Popen(chunk.ffmpeg_gen_cmd, stdout=PIPE, stderr=STDOUT)
+    pipe = subprocess.Popen(cmd, stdin=ffmpeg_gen_pipe.stdout, stdout=PIPE, stderr=STDOUT, shell=False, universal_newlines=True)
```
Comment (n9Mtq4) [inline Av1an/vmaf.py:74]:
```suggestion
    pipe = subprocess.Popen(cmd, stdin=ffmpeg_gen_pipe.stdout, stdout=PIPE, stderr=STDOUT, universal_newlines=True)
```

File: Av1an/chunk_queue.py (hunk 4 above)
Comment (Fuchs4) [reply Av1an/chunk_queue.py:153]: I am not sure about the first frame, but the last frame gets added by extra_split when using --extra_split option, at least for some of my videos. I added it as a quick fix and it is my first time working with github and pull requests, it was not supposed to be included in this pull request, and I noticed it too late. Sorry for that. We should probably look at extra_split for this issue.

Comment (master-of-zen) [reply Av1an/chunk_queue.py:153]: For sure extra splits should not exists in a way it is currently

File: Av1an/bar.py (hunk 2 above)
Comment (Fuchs4) [reply Av1an/bar.py]: compose_vvc creates a Tuple[Tuple[str]], like the other compose functions. I did this to be consistent with the other compose functions and to be prepared if we want to pipe from ffmpeg later. I do not know about vvc, but I guess it does not support reading from pipe (yet?). If you prefer using a Tuple[str] you also need to change compose_vvc.

Comment (Fuchs4) [reply Av1an/bar.py]: Sorry, you were right. I intended it to be a `Tuple[Tuple[str]]`, but looks like I missed a pair of brackets. It is all right with your change.

**Comment by master-of-zen** [comment]
@Fuchs4 Thanks! :+1: 
@n9Mtq4 Great work!  :hugs:

File: Av1an/chunk_queue.py (hunk 4 above)
Comment (n9Mtq4) [reply Av1an/chunk_queue.py]: Ok. We should take a look at extra splits. The reason I asked about this is that this is not the only place that the first and last frame is added to the split locations, so the other place might need changing as well.
